# Santa Fe Benchmark, SM

In [1]:
from pycgp.benchmarks.santafe import PARAMS, EV_PARAMS, X, santafe_cost_function
from pycgp.evolution import evolution
import numpy as np
import random

In [2]:
def run_experiment(params, ev_params, x):
    rstat = []
    for i in range(0, 20):
        print(i, end=', ')

        result = evolution(PARAMS, EV_PARAMS, X, None)

        rstat.append([santafe_cost_function(None, individual.execute(X)) for individual in result['final']])
    
    #print('Best fitness: {}'.format(np.min(stats)))
    #print('mean and std of fitness of last generation: {}, {}'.format(np.mean(stats), np.std(stats)))
    #print('Mean and std of best fitness: {}, {}'.format(np.mean(np.min(stats, axis=1)), np.std(np.min(stats, axis=1))))
    # best fitness, mean of last generation, std of last generation, mean of best individual, std of best individual
    results = [
        np.min(rstat), np.mean(rstat), np.std(rstat), np.mean(np.min(rstat, axis=1)), np.std(np.min(rstat, axis=1)) 
    ]
    print(results)
    return results
    

## Single mutation

In [3]:
from pycgp.mutation import single_mutation
from pycgp.gems import GemSM, MatchSMStrategy
EV_PARAMS['mutation']   = single_mutation
EV_PARAMS['match_strategy'] = MatchSMStrategy
EV_PARAMS['gem_type'] = GemSM
all_measurements = {}

### SM, 10 nodes

In [4]:
%%time
random.seed(1)

PARAMS['n_cols']        = 10
EV_PARAMS['gems']       = False
EV_PARAMS['j_box_size'] = 5

all_measurements['10,false'] = run_experiment(PARAMS, EV_PARAMS, X)

0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, [-75, -17.210000000000001, 13.344882914435781, -27.800000000000001, 14.451989482420752]
CPU times: user 1min 51s, sys: 297 ms, total: 1min 52s
Wall time: 1min 54s


### SM, 50 nodes

In [5]:
%%time
stats = []; random.seed(1);

PARAMS['n_cols']        = 50
EV_PARAMS['gems']       = False
EV_PARAMS['j_box_size'] = 5

all_measurements['50,false'] = run_experiment(PARAMS, EV_PARAMS, X)

0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, [-89, -57.020000000000003, 30.755805955949196, -66.549999999999997, 26.952690032722153]
CPU times: user 4min 45s, sys: 344 ms, total: 4min 45s
Wall time: 4min 53s


### SM, 100 nodes

In [6]:
%%time
stats = []; random.seed(1);

PARAMS['n_cols']        = 100
EV_PARAMS['gems']       = False
EV_PARAMS['j_box_size'] = 5

all_measurements['100,false'] = run_experiment(PARAMS, EV_PARAMS, X)

0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, [-89, -57.43, 27.761575963910978, -66.900000000000006, 23.649312886424418]
CPU times: user 7min 39s, sys: 188 ms, total: 7min 39s
Wall time: 7min 42s


### SM, 10 nodes, Gems

In [7]:
%%time
stats = []; random.seed(1);

PARAMS['n_cols']        = 10
EV_PARAMS['gems']       = True
EV_PARAMS['j_box_size'] = 5

all_measurements['10,true,5'] = run_experiment(PARAMS, EV_PARAMS, X)

0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, [-89, -18.870000000000001, 19.116827665698093, -32.850000000000001, 20.905202701719972]
CPU times: user 1min 54s, sys: 15.6 ms, total: 1min 54s
Wall time: 1min 55s


### SM, 50 nodes, Gems

In [ ]:
%%time
stats = []; random.seed(1);

PARAMS['n_cols']        = 50
EV_PARAMS['gems']       = True
EV_PARAMS['j_box_size'] = 5

all_measurements['50,false,5'] = run_experiment(PARAMS, EV_PARAMS, X)

0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 

### SM, 100 nodes, Gems

In [ ]:
%%time
stats = []; random.seed(1);

PARAMS['n_cols']        = 100
EV_PARAMS['gems']       = True
EV_PARAMS['j_box_size'] = 5

all_measurements['100,false,5'] = run_experiment(PARAMS, EV_PARAMS, X)

### SM, 10 nodes, Gems, jBox 10

In [ ]:
%%time
random.seed(1);

PARAMS['n_cols']        = 10
EV_PARAMS['gems']       = True
EV_PARAMS['j_box_size'] = 10

all_measurements['10,false,10'] = run_experiment(PARAMS, EV_PARAMS, X)

### SM, 50 nodes, Gems, jBox 10

In [ ]:
%%time
random.seed(1);

PARAMS['n_cols']        = 50
EV_PARAMS['gems']       = True
EV_PARAMS['j_box_size'] = 10

all_measurements['50,false,10'] = run_experiment(PARAMS, EV_PARAMS, X)

### SM, 100 nodes, Gems, jBox 10

In [ ]:
%%time
random.seed(1);

PARAMS['n_cols']        = 100
EV_PARAMS['gems']       = True
EV_PARAMS['j_box_size'] = 10

all_measurements['100,false,10'] = run_experiment(PARAMS, EV_PARAMS, X)

# Summary

In [ ]:
import pandas as pd
df = pd.DataFrame.from_dict(all_measurements, orient='index')
df.columns = ['best fitness', 'mean of last gen', 'std of last gen', 'mean of best individual', 'std of best indvidiual']
df

## Summary

**Without gems**

| nodes | time     | max fitness | last gen mean | last gen std | best ind mean | best ind std |
|-------|----------|------------:|------------  :|-------------:|--------------:|-------------:|
| 10    |  2min 44s| 75          | 23.4          | 13.34        | 27.8          | 14.45        |
| 50    |  6min 17s| 89          | 57.02         | 30.75        | 66.55         | 26.95        |
| 100   |  9min 22s| 89          | **57.43**     | 27.76        | **66.9**      | 23.65        |

**With gems, jBox size 5**

| nodes | time     | max fitness | last gen mean | last gen std | best ind mean | best ind std |
|-------|----------|------------:|------------  :|-------------:|--------------:|-------------:|
| 10    | 1min 52s | 89          | 18.87         | 19.92        | 32.85         | 20.91        |
| 50    | 4min 24s | 89          | 64.04         | 27.07        | 75.45         | 22.64        |
| 100   | 6min 0s  | 89          | 53.69         | 28.38        | 61.95         | 26.48        |

**With gems, jBox size 10**

| nodes | time     | max fitness | last gen mean | last gen std | best ind mean | best ind std |
|-------|----------|------------:|------------  :|-------------:|--------------:|-------------:|
| 10    | 1min 44s | 57          | 11.78         | 9.93         | 24.1          | 10.37        |
| 50    | 3min 26s | 89          | 54.47         | 33.08        | 65.5          | 30.07        |
| 100   | 9min 53s | 89          | 66.25         | 26.27        | 76.2          | 20.42        |